src/include/simeng/config/Expectations.hh

```diff
@@ -0,0 +1,403 @@
+
+  /** A templated struct to store a boolean value denoting whether a passed
+   * typename T belongs to one types represented in ExpectedType. */
+  template <typename T>
+  struct is_expected_type
+      : std::integral_constant<bool, std::is_same<bool, T>::value ||
+                                         std::is_same<double, T>::value ||
+                                         std::is_same<float, T>::value ||
+                                         std::is_same<int64_t, T>::value ||
+                                         std::is_same<std::string, T>::value ||
+                                         std::is_same<uint64_t, T>::value> {};
+
+  /** A templated function to allow for the creation of an `ExpectationNode`
+   * instance. The instance created is one with a value and a key. A default
+   * value, key, type, and a bool denoting whether the node is optional are
+   * provided to the underlying constructor. */
+  template <typename T>
+  static ExpectationNode createExpectation(T defaultValue, std::string key,
+                                           bool optional = false) {
+    // Ensure templated type is of an expected type
+    static_assert(is_expected_type<T>::value &&
+                  "[SimEng:ModelConfig] Unexpected type given to "
+                  "ExpectationNode::create");
+    DataTypeVariant defValVariant = defaultValue;
+    ExpectedType type = static_cast<ExpectedType>(defValVariant.index());
+    ExpectationNode node = ExpectationNode(key, type, optional);
+    node.setDefaultValue(defValVariant);
+    return node;
+  }
+
+  /** A templated function to allow for the creation of an `ExpectationNode`
+   * instance with a key but no associated value. A key and a bool denoting
+   * whether the node is optional are provided to the underlying constructor. */
+  static ExpectationNode createExpectation(std::string key,
+                                           bool optional = false) {
+    ExpectationNode node =
+        ExpectationNode(key, ExpectedType::Valueless, optional);
+    return node;
+  }
+
+  /** Constructor for ExpectationNode instances. */
+  ExpectationNode(std::string key, ExpectedType type, bool optional)
+      : nodeKey_(key), type_(type), isOptional_(optional) {
+    if (nodeKey_ == wildcard) isWild_ = true;
+  }
+
+  ExpectationNode(){};
+
+  ~ExpectationNode(){};
+
+  /** A getter function to retrieve the key of a node. */
+  std::string getKey() const { return nodeKey_; }
+
+  /** A getter function to retrieve the held default value of a node. */
+  template <typename T>
+  T getDefault() const {
+    return getByType<T>(defaultValue_);
+  }
+
+  /** A getter function to retrieve the value type of a node. */
+  ExpectedType getType() const { return type_; }
+
+  /** A getter function to retrieve the child ExpectationNode instances of this
+   * node. */
+  std::vector<ExpectationNode> getChildren() const { return nodeChildren_; }
```

Comment:
You can't return a const here as we call `validateConfigNode` via the children nodes which cannot be qualified as a const given we possibly alter the passed node with default values. We could pass a reference but then would have to look the const qualifier of the getter function

Reply:
I may be missing something, but `validateConfigNode` and `validateConfigNodeWithType` don't seem to change the ExpectationNode itself, but rather just the ryaml::NodeRef passed into these functions?

If this isn't the case though - I feel as though the latter would be preferable (removing const but returning a reference to the children) to avoid multiple copies + creation of vectors?

Reply:
I'll change to be a ref

Reply:
If these functions do not alter any member variables of the class they belong to then they should be able to be made `const` - even if they change any argument passed in

Reply:
See edited comments